drivers/android/raw_list.rs

```diff
@@ -0,0 +1,288 @@
+}
+
+struct ListEntry<T: ?Sized> {
+    next: Option<NonNull<T>>,
+    prev: Option<NonNull<T>>,
+    inserted: AtomicBool,
+}
+
+impl<T: ?Sized> ListEntry<T> {
+    fn new() -> Self {
+        Self {
+            next: None,
+            prev: None,
+            inserted: AtomicBool::new(false),
+        }
+    }
+
+    fn acquire_for_insertion(&mut self) -> bool {
+        self.inserted
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_ok()
+    }
+
+    fn release_after_removal(&mut self) {
+        self.inserted.store(false, Ordering::Release);
+    }
+}
+
+pub struct RawList<G: GetLinks> {
+    head: Option<NonNull<G::EntryType>>,
+}
+
+impl<G: GetLinks> RawList<G> {
+    pub fn new() -> Self {
+        Self { head: None }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.head.is_none()
+    }
+
+    fn insert_after_priv(
+        &mut self,
+        existing: &G::EntryType,
+        new_entry: &mut ListEntry<G::EntryType>,
+        new_ptr: Option<NonNull<G::EntryType>>,
+    ) -> bool {
+        if !new_entry.acquire_for_insertion() {
+            // Nothing to do if already inserted.
+            return false;
+        }
+
+        {
+            let existing_links = unsafe { &mut *G::get_links(existing).0.get() };
+            new_entry.next = existing_links.next;
+            existing_links.next = new_ptr;
+        }
+
+        new_entry.prev = Some(NonNull::from(existing));
+        let next_links = unsafe { &mut *G::get_links(new_entry.next.unwrap().as_ref()).0.get() };
+        next_links.prev = new_ptr;
+        true
+    }
+
+    pub fn insert_after(&mut self, existing: &G::EntryType, new: &G::EntryType) -> bool {
+        let new_entry = unsafe { &mut *G::get_links(new).0.get() };
+        self.insert_after_priv(existing, new_entry, Some(NonNull::from(new)))
+    }
+
+    fn push_back_internal(&mut self, new: &G::EntryType) -> bool {
+        let new_entry = unsafe { &mut *G::get_links(new).0.get() };
+        let new_ptr = Some(NonNull::from(new));
+        match self.back() {
+            Some(back) => self.insert_after_priv(unsafe { back.as_ref() }, new_entry, new_ptr),
+            None => {
+                if !new_entry.acquire_for_insertion() {
+                    // Nothing to do if already inserted.
+                    return false;
+                }
+                self.head = new_ptr;
+                new_entry.next = new_ptr;
+                new_entry.prev = new_ptr;
+                true
+            }
+        }
+    }
+
+    pub unsafe fn push_back(&mut self, new: &G::EntryType) -> bool {
+        self.push_back_internal(new)
+    }
+
+    fn remove_internal(&mut self, data: &G::EntryType) -> bool {
+        let links = unsafe { &mut *G::get_links(data).0.get() };
+        let next = if let Some(next) = links.next {
+            next
+        } else {
+            // Nothing to do if the entry is not on the list.
+            return false;
+        };
```

Comment:
Hopefully the `unless`-like proposal ends up being accepted in Rust so that we can do this pattern easily :-)

(Nothing to do here)